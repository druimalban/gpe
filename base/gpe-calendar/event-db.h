/*
 * Copyright (C) 2002 Philip Blundell <philb@gnu.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 */

#ifndef EVENT_DB_H
#define EVENT_DB_H

#include <glib.h>
#include <time.h>

typedef enum
{
  RECUR_NONE,
  RECUR_DAILY,
  RECUR_WEEKLY,
  RECUR_MONTHLY,
  RECUR_YEARLY
} recur_t;

typedef struct event_details_s
{
  char *summary;
  char *description;
} *event_details_t;

typedef struct event_s
{
  unsigned long uid;

  time_t start;
  unsigned long duration;	/* 0 == instantaneous */
  unsigned long alarm;		/* -1 == none */
  
  struct
  {
    recur_t type;

    unsigned int count;
    unsigned int daymask;	/* bit 0 = Mon, bit 1 = Tue, etc */

    time_t end;			/* 0 == perpetual */
  } recur;

  event_details_t details;
  
  gboolean mark;
} *event_t;

extern gboolean event_db_start (void);
extern gboolean event_db_stop (void);

extern gboolean event_db_add (event_t);
extern gboolean event_db_remove (event_t);

extern event_t event_db_new (void);
extern void event_db_destroy (event_t);

extern event_details_t event_db_alloc_details (event_t);
extern event_details_t event_db_get_details (event_t);
extern void event_db_forget_details (event_t);

extern GSList *event_db_list_for_period (time_t start, time_t end);

#endif
