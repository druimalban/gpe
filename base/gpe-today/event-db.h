/*
 * Copyright (C) 2002 Philip Blundell <philb@gnu.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 */

#ifndef EVENT_DB_H
#define EVENT_DB_H

#include <glib.h>
#include <time.h>
#include <stdlib.h>

typedef enum
{
  RECUR_NONE,
  RECUR_DAILY,
  RECUR_WEEKLY,
  RECUR_MONTHLY,
  RECUR_YEARLY
} recur_type_t;

typedef struct event_details_s
{
  guint ref;

  unsigned long sequence;
  time_t modified;

  char *summary;
  char *description;
} *event_details_t;

#define FLAG_UNTIMED   (1 << 0)
#define FLAG_ALARM     (1 << 1)
#define FLAG_TENTATIVE (1 << 2)

struct calendar_time_s
{
  u_int16_t year;
  u_int8_t month, day;
  u_int8_t hour, minute;
};

typedef time_t calendar_time_t;

typedef struct recur_s
{
  recur_type_t type;
  
  unsigned int count;
  unsigned int increment;
  unsigned long daymask;	/* bit 0 = Mon, bit 1 = Tue, etc */
  
  time_t end;			/* 0 == perpetual */
} *recur_t;

typedef struct event_s
{
  unsigned long uid;

  calendar_time_t start;
  unsigned long duration;	/* 0 == instantaneous */
  unsigned long alarm;		/* seconds before event */
  unsigned long flags;

  recur_t recur;
  
  event_details_t details;
  gboolean mark;
} *event_t;

extern gboolean event_db_start (void);
extern gboolean event_db_stop (void);

extern gboolean event_db_add (event_t);
extern gboolean event_db_remove (event_t);

extern gboolean event_db_flush (event_t ev);

extern event_t event_db_new (void);
extern void event_db_destroy (event_t);

extern event_details_t event_db_alloc_details (event_t);
extern event_details_t event_db_get_details (event_t);
extern void event_db_forget_details (event_t);

extern GSList *event_db_list_for_period (time_t start, time_t end);
extern GSList *event_db_list_for_future (time_t start, guint max);
extern GSList *event_db_untimed_list_for_period (time_t start, time_t end, gboolean yes);
extern void event_db_list_destroy (GSList *);

extern recur_t event_db_get_recurrence (event_t ev);

#endif
