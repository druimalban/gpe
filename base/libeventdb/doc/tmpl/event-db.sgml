<!-- ##### SECTION Title ##### -->
event-db

<!-- ##### SECTION Short_Description ##### -->


<!-- ##### SECTION Long_Description ##### -->
<para>

</para>

<!-- ##### SECTION See_Also ##### -->
<para>

</para>

<!-- ##### SECTION Stability_Level ##### -->


<!-- ##### MACRO SECONDS_IN_DAY ##### -->
<para>

</para>



<!-- ##### MACRO MON ##### -->
<para>

</para>



<!-- ##### MACRO TUE ##### -->
<para>

</para>



<!-- ##### MACRO WED ##### -->
<para>

</para>



<!-- ##### MACRO THU ##### -->
<para>

</para>



<!-- ##### MACRO FRI ##### -->
<para>

</para>



<!-- ##### MACRO SAT ##### -->
<para>

</para>



<!-- ##### MACRO SUN ##### -->
<para>

</para>



<!-- ##### ENUM recur_type_t ##### -->
<para>

</para>

@RECUR_NONE: 
@RECUR_DAILY: 
@RECUR_WEEKLY: 
@RECUR_MONTHLY: 
@RECUR_YEARLY: 

<!-- ##### MACRO FLAG_UNTIMED ##### -->
<para>

</para>



<!-- ##### MACRO FLAG_ALARM ##### -->
<para>

</para>



<!-- ##### MACRO FLAG_TENTATIVE ##### -->
<para>

</para>



<!-- ##### MACRO FLAG_CLONE ##### -->
<para>

</para>



<!-- ##### MACRO FLAG_RECUR ##### -->
<para>

</para>



<!-- ##### STRUCT calendar_time_s ##### -->
<para>

</para>

@date: 
@time: 

<!-- ##### TYPEDEF calendar_time_t ##### -->
<para>

</para>


<!-- ##### MACRO EVENT_DB_USE_MEMCHUNK ##### -->
<para>

</para>



<!-- ##### MACRO event_db__alloc_event ##### -->
<para>

</para>



<!-- ##### MACRO event_db__alloc_recur ##### -->
<para>

</para>



<!-- ##### MACRO event_db__free_event ##### -->
<para>

</para>

@_x: 


<!-- ##### MACRO event_db__free_recur ##### -->
<para>

</para>

@_x: 


<!-- ##### FUNCTION event_db_start ##### -->
<para>

</para>

@Returns: 


<!-- ##### FUNCTION event_db_refresh ##### -->
<para>

</para>

@Returns: 


<!-- ##### FUNCTION event_db_stop ##### -->
<para>

</para>

@Returns: 


<!-- ##### FUNCTION event_db_add ##### -->
<para>

</para>

@Param1: 
@Returns: 


<!-- ##### FUNCTION event_db_remove ##### -->
<para>

</para>

@Param1: 
@Returns: 


<!-- ##### FUNCTION event_db_flush ##### -->
<para>

</para>

@ev: 
@Returns: 


<!-- ##### FUNCTION event_db_clone ##### -->
<para>

</para>

@Param1: 
@Returns: 


<!-- ##### FUNCTION event_db_new ##### -->
<para>

</para>

@Returns: 


<!-- ##### FUNCTION event_db_find_by_uid ##### -->
<para>

</para>

@uid: 
@Returns: 


<!-- ##### FUNCTION event_db_destroy ##### -->
<para>

</para>

@Param1: 


<!-- ##### FUNCTION event_db_alloc_details ##### -->
<para>

</para>

@Param1: 
@Returns: 


<!-- ##### FUNCTION event_db_get_details ##### -->
<para>

</para>

@Param1: 
@Returns: 


<!-- ##### FUNCTION event_db_forget_details ##### -->
<para>

</para>

@Param1: 


<!-- ##### FUNCTION event_db_list_for_period ##### -->
<para>

</para>

@start: 
@end: 
@Returns: 


<!-- ##### FUNCTION event_db_list_alarms_for_period ##### -->
<para>

</para>

@start: 
@end: 
@Returns: 


<!-- ##### FUNCTION event_db_list_for_future ##### -->
<para>

</para>

@start: 
@max: 
@Returns: 


<!-- ##### FUNCTION event_db_untimed_list_for_period ##### -->
<para>

</para>

@start: 
@end: 
@yes: 
@Returns: 


<!-- ##### FUNCTION event_db_list_destroy ##### -->
<para>

</para>

@Param1: 


<!-- ##### FUNCTION event_db_get_recurrence ##### -->
<para>

</para>

@ev: 
@Returns: 


