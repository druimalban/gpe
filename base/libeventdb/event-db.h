/* event-db.h: Event Database Interface.
 * Copyright (C) 2002, 2006 Philip Blundell <philb@gnu.org>
 *               2006, Florian Boor <florian@kernelconcepts.de>
 * Copyright (C) 2006 Neal H. Walfield <neal@walfield.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 */

#ifndef EVENT_DB_H
#define EVENT_DB_H

#include <glib-object.h>
#include <glib.h>
#include <time.h>

#include "event-cal.h"
#include "event.h"

#define ERROR_DOMAIN() g_quark_from_static_string ("libeventdb")
/* If an error occurs and the passed GError is NULL, we propagate
   error by emitting a signal on EDB.  This must be used instead of
   g_set_error.  */
#define SIGNAL_ERROR(edb, gerror, args , ...) \
  do \
    { \
      char *__e = g_strdup_printf (args, ##__VA_ARGS__); \
      if (gerror) \
        g_set_error (gerror, ERROR_DOMAIN (), 0, \
                     "%s: %s", __func__, __e); \
      else \
        { \
          char *buffer = g_strdup_printf ("%s: %s", __func__, __e); \
          g_signal_emit \
            (edb, EVENT_DB_GET_CLASS (edb)->error_signal, 0, buffer); \
          g_free (buffer); \
        } \
      g_free (__e); \
    } \
  while (0)
#define SIGNAL_ERROR_GERROR(edb, dest, src) \
  do \
    { \
      if ((dest)) \
        g_propagate_error ((dest), (src)); \
      else \
        { \
          char *buffer = g_strdup_printf ("%s:%s", __func__, (src)->message); \
          g_error_free ((src)); \
          g_signal_emit \
            ((edb), EVENT_DB_GET_CLASS (edb)->error_signal, 0, buffer); \
          g_free (buffer); \
        } \
    } \
  while (0)

/* Enumerate the events in the event database EDB which MAY occur
   between (PERIOD_START and PERIOD_END].  If ALARMS is true, returns
   those events which have an alarm which MAY go off between
   PERIOD_START and PERIOD_END.  Calls CB on each event until CB
   returns a non-zero value.  A reference to EV is allocated and the
   callback function must consume it.  */
typedef void (*events_enumerate_t) (EventDB *edb,
				    time_t period_start, time_t period_end,
				    gboolean alarms,
				    int (*cb) (EventSource *ev),
				    GError **error);
/* Returns the event with the event id EVENTID or NULL if there is no
   such event.  */
typedef int (*eventid_to_uid_t) (EventDB *edb, const char *eventid,
				 GError **error);
/* Set the default calendar to EC.  */
typedef void (*set_default_calendar_t) (EventDB *edb, EventCalendar *ec,
					GError **error);

/* Create a new event on backing store.  EV is uninitialized; save the
   allocated UID in EV->UID.  */
typedef gboolean (*event_new_t) (EventSource *ev, GError **error);
/* Load event EV's basic data (i.e. not necessarily its details) from
   backing store.  EV->UID is valid.  Returns FALSE if no such event
   exists.  */
typedef gboolean (*event_load_t) (EventSource *ev, GError **error);
/* Load the event's details from backing store.  event_load has
   already been called successfully.  */
typedef void (*event_load_details_t) (EventSource *ev, GError **error);
/* Flush the event to backing store.  If an error occurs, FALSE should
   be returned and an error may be reported in *ERR which the client
   must free using g_free.  */
typedef gboolean (*event_flush_t) (EventSource *ev, GError **error);
/* Mark the event EV as removed (i.e. deleted).  */
typedef void (*event_remove_t) (EventSource *ev, GError **error);

/* List the events which have alarms which have not yet been
   acknowledged.  */
typedef GSList *(*list_unacknowledged_alarms_t) (EventDB *edb, GError **error);
/* Mark EV's alarm as having fired but yet not been acknowledged.  */
typedef void (*event_mark_unacknowledged_t) (EventSource *ev, GError **error);
/* Mark EV's alarm as having been acknowledged.  */
typedef void (*event_mark_acknowledged_t) (EventSource *ev, GError **error);
/* Mark all alarms having fired through T as having been
   acknowledged.  */
typedef void (*acknowledge_alarms_through_t) (EventDB *edb, time_t t,
					      GError **error);

/* Create a new calendar on the backing store using EC as the
   template.  Must initialize EC->UID to an identifier unique to the
   backing store.  */
typedef void (*event_calendar_new_t) (EventCalendar *ec, GError **error);
/* Flush the calendar to backing store.  */
typedef void (*event_calendar_flush_t) (EventCalendar *ec, GError **error);
/* Delete the calendar from the backing store.  The generic framework
   has already taken care that no dangling references are left
   around.  */
typedef void (*event_calendar_delete_t) (EventCalendar *ec, GError **error);
/* List the events in the calendar EC.  If MODIFIED_AFTER is non-zero,
   limit to those events which have been modified on or after
   MODIFIED_AFTER.  If MODIFIED_BEFORE is non-zero, limit to those
   events which have been modified on or before MODIFIED_BEFORE.  */
typedef GSList *(*event_calendar_list_events_t) (EventCalendar *ec,
						 time_t modified_after,
						 time_t modified_before,
						 GError **error);
/* List the deleted events in the calendar EC.  NB: Deleted events
   have EV->DEAD set to true.  */
typedef GSList *(*event_calendar_list_deleted_t) (EventCalendar *ec,
						  GError **error);
/* Flush any deleted events from EC.  */
typedef void (*event_calendar_flush_deleted_t) (EventCalendar *ec,
						GError **error);

typedef struct
{
  GObjectClass gobject_class;

  events_enumerate_t events_enumerate;
  eventid_to_uid_t eventid_to_uid;
  set_default_calendar_t set_default_calendar;

  event_new_t event_new;
  event_load_t event_load;
  event_load_details_t event_load_details;
  event_flush_t event_flush;
  event_remove_t event_remove;

  list_unacknowledged_alarms_t list_unacknowledged_alarms;
  event_mark_unacknowledged_t event_mark_unacknowledged;
  event_mark_acknowledged_t event_mark_acknowledged;
  acknowledge_alarms_through_t acknowledge_alarms_through;

  event_calendar_new_t event_calendar_new;
  event_calendar_flush_t event_calendar_flush;
  event_calendar_delete_t event_calendar_delete;
  event_calendar_list_events_t event_calendar_list_events;
  event_calendar_list_deleted_t event_calendar_list_deleted;
  event_calendar_flush_deleted_t event_calendar_flush_deleted;

  /* Signals.  */
  guint error_signal;
  EventDBError error;

  guint calendar_new_signal;
  EventCalendarNew calendar_new;
  guint calendar_deleted_signal;
  EventCalendarDeleted calendar_deleted;
  guint calendar_reparented_signal;
  EventCalendarReparented calendar_reparented;
  guint calendar_changed_signal;
  EventCalendarChanged calendar_changed;
  guint calendar_modified_signal;
  EventCalendarModified calendar_modified;
  
  guint event_new_signal;
  EventNew event_new_func;
  guint event_removed_signal;
  EventRemoved event_removed;
  guint event_modified_signal;
  EventModified event_modified;

  guint alarm_fired_signal;
  EventDBAlarmFiredFunc alarm_fired;
} EventDBClass;

struct _EventDB
{
  GObject object;

  GHashTable *events;
  guint default_calendar;
  GSList *calendars;

  /* A list of events that need to be flushed to backing store.  */
  GSList *laundry_list;
  /* The idle source.  */
  guint laundry_buzzer;

  /* A list of events which have no user references.  */
  GSList *cache_list;
  /* Number of seconds to leave unreferenced items in memory before
     deallocating them.  */
#ifndef CACHE_EXPIRE
#define CACHE_EXPIRE 120
#endif
  /* A timer source to periodically flush the cache.  */
  guint cache_buzzer;

  /* The list of events with upcoming alarms.  We hold a reference to
     each.  */
  GSList *upcoming_alarms;
  /* EVENTS contains alarms until this point in time.  */
  time_t period_end;

  /* The alarm source.  */
  guint alarm;

  /* The point through which alarms have been fired (when an alarm
     fires, it is entered into the alarms_unacknowledged table and
     only removed once it has been acknowledged).  */
  time_t alarms_fired_through;
};

/* E is an Event or an EventCalendar and is dirty (i.e. needs to be
   written to disk).  Add it to the laundry list and do it when we are
   idle.  */
extern void add_to_laundry_pile (GObject *e)
     __attribute__ ((visibility ("hidden")));

/* Invoked by a timeout source when either an alarm should go off or
   when we need to look for additional upcoming events.  */
extern gboolean buzzer (gpointer data)
     __attribute__ ((visibility ("hidden")));

/* All EventSources are created with a toggle reference.  When the
   last user reference is dropped, this function is invoked and we
   grab a real reference and stick the event on the EDB->CACHE_LIST
   list.  If no user reference is taken before CACHE_EXPIRE seconds,
   the event is freed.  */
extern void event_source_toggle_ref_notify (gpointer data,
					    GObject *object,
					    gboolean is_last_ref)
     __attribute__ ((visibility ("hidden")));

#endif
