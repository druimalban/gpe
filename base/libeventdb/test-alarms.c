/* test-alarms.c - Test the event_db_list_alarms_for_period.
   Copyright (C) 2006 Neal H. Walfield <neal@walfield.org>

   This file is part of GPE.

   GPE is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   GPE is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
   or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
   License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111, USA. */

int do_test (int argc, char *argv[]);

#include "test-skeleton.h"

#include <stdio.h>
#include <error.h>
#include "gpe/event-db.h"

static int fail;

#define TIME_TO_STRING(t, buffer) \
  ({ \
    struct tm _tm; \
    time_t __t = t; \
    localtime_r (&__t, &_tm); \
    strftime (buffer, sizeof (buffer), "%F %T", &_tm); \
   })

int
do_test (int argc, char *argv[])
{
  char *file;
  int fd;
  /* Tue Jun 20 07:00:00 UTC 2006.  */
  time_t start = 1150786800;
  setenv ("TZ", "UTC", 1);
  tzset ();

  /* Initialize the g_object system.  */
  g_type_init ();

  fd = create_temp_file (".", &file);
  /* event_db_new will open it itself.  */
  close (fd);

  GError *err = NULL;
  EventDB *edb = event_db_new (file, &err);
  if (! edb)
    error (1, 0, "evend_db_new: %s", err->message);

  void edb_error (EventDB *edb, const char *error)
    {
      puts (error);
    }
  g_signal_connect (G_OBJECT (edb),
		    "error", G_CALLBACK (edb_error), NULL);

  /* One hour.  */
#define INTERVAL 60 * 60
  /* Ten minutes.  */
#define ALARM 10 * 60

  /* One time event -- no alarm.  */
  Event *ev = event_new (edb, NULL, NULL, NULL);
  event_set_summary (ev, "no alarm", NULL);
  event_set_recurrence_start (ev, start, NULL);
  event_set_duration (ev, 60, NULL);
  g_object_unref (ev);

  /* One time event -- alarm.  */
  ev = event_new (edb, NULL, NULL, NULL);
  event_set_summary (ev, "once at 7:10", NULL);
  event_set_recurrence_start (ev, start + ALARM, NULL);
  event_set_duration (ev, 60, NULL);
  event_set_alarm (ev, ALARM, NULL);
  g_object_unref (ev);

  /* Daily event -- alarm.  */
  ev = event_new (edb, NULL, NULL, NULL);
  event_set_summary (ev, "daily at 7, twice", NULL);
  event_set_recurrence_start (ev, start, NULL);
  event_set_duration (ev, 1, NULL);
  event_set_alarm (ev, ALARM, NULL);
  event_set_recurrence_type (ev, RECUR_DAILY, NULL);
  event_set_recurrence_count (ev, 2, NULL);
  g_object_unref (ev);

  int i;
  for (i = -5; i < 3 * 24; i ++)
    {
      time_t s = start + i * 60 * 60;
      time_t e = s + 60 * 60 - 1;

      printf ("Next alarm after ");
      char buffer[200];
      TIME_TO_STRING (s, buffer);
      printf ("%s", buffer);

      Event *ev = event_db_next_alarm (edb, s, NULL);
      if (! ev)
	printf (" none\n");
      else
	{
	  printf (" %s at ", event_get_summary (ev, NULL));
	  TIME_TO_STRING (event_get_start (ev) - event_get_alarm (ev),
			  buffer);
	  printf ("%s\n", buffer);

	  g_object_unref (ev);
	}

      GSList *list
	= event_db_list_alarms_for_period (edb, s, e, NULL);

      TIME_TO_STRING (s, buffer);
      fputs (buffer, stdout);
      fputs (" until ", stdout);
      TIME_TO_STRING (e, buffer);
      puts (buffer);

      GSList *l;
      list = g_slist_sort (list, event_compare_func);
      for (l = list; l; l = l->next)
	{
	  Event *ev = EVENT (l->data);
	  TIME_TO_STRING (event_get_start (ev), buffer);
	  printf ("%s: %s\n", buffer, event_get_summary (ev, NULL));
	}

      event_list_unref (list);
    }

  return fail;
}

