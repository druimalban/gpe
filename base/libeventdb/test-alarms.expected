Next alarm after 2006-06-20 02:00:00 daily at 7, twice at 2006-06-20 06:50:00
2006-06-20 02:00:00 until 2006-06-20 02:59:59
Next alarm after 2006-06-20 03:00:00 daily at 7, twice at 2006-06-20 06:50:00
2006-06-20 03:00:00 until 2006-06-20 03:59:59
Next alarm after 2006-06-20 04:00:00 daily at 7, twice at 2006-06-20 06:50:00
2006-06-20 04:00:00 until 2006-06-20 04:59:59
Next alarm after 2006-06-20 05:00:00 daily at 7, twice at 2006-06-20 06:50:00
2006-06-20 05:00:00 until 2006-06-20 05:59:59
Next alarm after 2006-06-20 06:00:00 daily at 7, twice at 2006-06-20 06:50:00
2006-06-20 06:00:00 until 2006-06-20 06:59:59
2006-06-20 07:00:00: daily at 7, twice
Next alarm after 2006-06-20 07:00:00 once at 7:10 at 2006-06-20 07:00:00
2006-06-20 07:00:00 until 2006-06-20 07:59:59
2006-06-20 07:10:00: once at 7:10
Next alarm after 2006-06-20 08:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 08:00:00 until 2006-06-20 08:59:59
Next alarm after 2006-06-20 09:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 09:00:00 until 2006-06-20 09:59:59
Next alarm after 2006-06-20 10:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 10:00:00 until 2006-06-20 10:59:59
Next alarm after 2006-06-20 11:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 11:00:00 until 2006-06-20 11:59:59
Next alarm after 2006-06-20 12:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 12:00:00 until 2006-06-20 12:59:59
Next alarm after 2006-06-20 13:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 13:00:00 until 2006-06-20 13:59:59
Next alarm after 2006-06-20 14:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 14:00:00 until 2006-06-20 14:59:59
Next alarm after 2006-06-20 15:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 15:00:00 until 2006-06-20 15:59:59
Next alarm after 2006-06-20 16:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 16:00:00 until 2006-06-20 16:59:59
Next alarm after 2006-06-20 17:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 17:00:00 until 2006-06-20 17:59:59
Next alarm after 2006-06-20 18:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 18:00:00 until 2006-06-20 18:59:59
Next alarm after 2006-06-20 19:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 19:00:00 until 2006-06-20 19:59:59
Next alarm after 2006-06-20 20:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 20:00:00 until 2006-06-20 20:59:59
Next alarm after 2006-06-20 21:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 21:00:00 until 2006-06-20 21:59:59
Next alarm after 2006-06-20 22:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 22:00:00 until 2006-06-20 22:59:59
Next alarm after 2006-06-20 23:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-20 23:00:00 until 2006-06-20 23:59:59
Next alarm after 2006-06-21 00:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-21 00:00:00 until 2006-06-21 00:59:59
Next alarm after 2006-06-21 01:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-21 01:00:00 until 2006-06-21 01:59:59
Next alarm after 2006-06-21 02:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-21 02:00:00 until 2006-06-21 02:59:59
Next alarm after 2006-06-21 03:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-21 03:00:00 until 2006-06-21 03:59:59
Next alarm after 2006-06-21 04:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-21 04:00:00 until 2006-06-21 04:59:59
Next alarm after 2006-06-21 05:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-21 05:00:00 until 2006-06-21 05:59:59
Next alarm after 2006-06-21 06:00:00 daily at 7, twice at 2006-06-21 06:50:00
2006-06-21 06:00:00 until 2006-06-21 06:59:59
2006-06-21 07:00:00: daily at 7, twice
Next alarm after 2006-06-21 07:00:00 none
2006-06-21 07:00:00 until 2006-06-21 07:59:59
Next alarm after 2006-06-21 08:00:00 none
2006-06-21 08:00:00 until 2006-06-21 08:59:59
Next alarm after 2006-06-21 09:00:00 none
2006-06-21 09:00:00 until 2006-06-21 09:59:59
Next alarm after 2006-06-21 10:00:00 none
2006-06-21 10:00:00 until 2006-06-21 10:59:59
Next alarm after 2006-06-21 11:00:00 none
2006-06-21 11:00:00 until 2006-06-21 11:59:59
Next alarm after 2006-06-21 12:00:00 none
2006-06-21 12:00:00 until 2006-06-21 12:59:59
Next alarm after 2006-06-21 13:00:00 none
2006-06-21 13:00:00 until 2006-06-21 13:59:59
Next alarm after 2006-06-21 14:00:00 none
2006-06-21 14:00:00 until 2006-06-21 14:59:59
Next alarm after 2006-06-21 15:00:00 none
2006-06-21 15:00:00 until 2006-06-21 15:59:59
Next alarm after 2006-06-21 16:00:00 none
2006-06-21 16:00:00 until 2006-06-21 16:59:59
Next alarm after 2006-06-21 17:00:00 none
2006-06-21 17:00:00 until 2006-06-21 17:59:59
Next alarm after 2006-06-21 18:00:00 none
2006-06-21 18:00:00 until 2006-06-21 18:59:59
Next alarm after 2006-06-21 19:00:00 none
2006-06-21 19:00:00 until 2006-06-21 19:59:59
Next alarm after 2006-06-21 20:00:00 none
2006-06-21 20:00:00 until 2006-06-21 20:59:59
Next alarm after 2006-06-21 21:00:00 none
2006-06-21 21:00:00 until 2006-06-21 21:59:59
Next alarm after 2006-06-21 22:00:00 none
2006-06-21 22:00:00 until 2006-06-21 22:59:59
Next alarm after 2006-06-21 23:00:00 none
2006-06-21 23:00:00 until 2006-06-21 23:59:59
Next alarm after 2006-06-22 00:00:00 none
2006-06-22 00:00:00 until 2006-06-22 00:59:59
Next alarm after 2006-06-22 01:00:00 none
2006-06-22 01:00:00 until 2006-06-22 01:59:59
Next alarm after 2006-06-22 02:00:00 none
2006-06-22 02:00:00 until 2006-06-22 02:59:59
Next alarm after 2006-06-22 03:00:00 none
2006-06-22 03:00:00 until 2006-06-22 03:59:59
Next alarm after 2006-06-22 04:00:00 none
2006-06-22 04:00:00 until 2006-06-22 04:59:59
Next alarm after 2006-06-22 05:00:00 none
2006-06-22 05:00:00 until 2006-06-22 05:59:59
Next alarm after 2006-06-22 06:00:00 none
2006-06-22 06:00:00 until 2006-06-22 06:59:59
Next alarm after 2006-06-22 07:00:00 none
2006-06-22 07:00:00 until 2006-06-22 07:59:59
Next alarm after 2006-06-22 08:00:00 none
2006-06-22 08:00:00 until 2006-06-22 08:59:59
Next alarm after 2006-06-22 09:00:00 none
2006-06-22 09:00:00 until 2006-06-22 09:59:59
Next alarm after 2006-06-22 10:00:00 none
2006-06-22 10:00:00 until 2006-06-22 10:59:59
Next alarm after 2006-06-22 11:00:00 none
2006-06-22 11:00:00 until 2006-06-22 11:59:59
Next alarm after 2006-06-22 12:00:00 none
2006-06-22 12:00:00 until 2006-06-22 12:59:59
Next alarm after 2006-06-22 13:00:00 none
2006-06-22 13:00:00 until 2006-06-22 13:59:59
Next alarm after 2006-06-22 14:00:00 none
2006-06-22 14:00:00 until 2006-06-22 14:59:59
Next alarm after 2006-06-22 15:00:00 none
2006-06-22 15:00:00 until 2006-06-22 15:59:59
Next alarm after 2006-06-22 16:00:00 none
2006-06-22 16:00:00 until 2006-06-22 16:59:59
Next alarm after 2006-06-22 17:00:00 none
2006-06-22 17:00:00 until 2006-06-22 17:59:59
Next alarm after 2006-06-22 18:00:00 none
2006-06-22 18:00:00 until 2006-06-22 18:59:59
Next alarm after 2006-06-22 19:00:00 none
2006-06-22 19:00:00 until 2006-06-22 19:59:59
Next alarm after 2006-06-22 20:00:00 none
2006-06-22 20:00:00 until 2006-06-22 20:59:59
Next alarm after 2006-06-22 21:00:00 none
2006-06-22 21:00:00 until 2006-06-22 21:59:59
Next alarm after 2006-06-22 22:00:00 none
2006-06-22 22:00:00 until 2006-06-22 22:59:59
Next alarm after 2006-06-22 23:00:00 none
2006-06-22 23:00:00 until 2006-06-22 23:59:59
Next alarm after 2006-06-23 00:00:00 none
2006-06-23 00:00:00 until 2006-06-23 00:59:59
Next alarm after 2006-06-23 01:00:00 none
2006-06-23 01:00:00 until 2006-06-23 01:59:59
Next alarm after 2006-06-23 02:00:00 none
2006-06-23 02:00:00 until 2006-06-23 02:59:59
Next alarm after 2006-06-23 03:00:00 none
2006-06-23 03:00:00 until 2006-06-23 03:59:59
Next alarm after 2006-06-23 04:00:00 none
2006-06-23 04:00:00 until 2006-06-23 04:59:59
Next alarm after 2006-06-23 05:00:00 none
2006-06-23 05:00:00 until 2006-06-23 05:59:59
Next alarm after 2006-06-23 06:00:00 none
2006-06-23 06:00:00 until 2006-06-23 06:59:59
